docs/proposals/flexvolume-api.md

```diff
@@ -0,0 +1,177 @@
+# Flex Volume API Proposal
+## Abstract
+The goal of this proposal is to extend and enhance the existing flex volume storage plugin model (#13840). It bridges the gap with existing in-tree plugins and defines a stable api.
+
+## Background:
+The existing flex volume plugin enables administrators/vendors to create out-of-tree volume plugins, i.e., plugins which do not reside in Kubernetes project. But, it has the following gaps:
+* It lacks support for the latest enhancements to in-tree plugins like dynamic provisioning and centralized attach-detach controller
+* It does not define a stable driver api for administrators/vendors to work with.
+* The current flex volume plugin deployment model is not suitable for environments like GCE & CoreOS, where access to the root file system is restricted. And in other environments the overall user experience is poor, as it requires manual installation of the plugin on every node and restart of kubelets & controller manager. This issue is covered in detail in another PR #32550.
+
+### Goals:
+The goals of this proposal are 
+* Define new stable driver API.
+* Add support for dynamic provisioning and centralized attach-detach controller and
+* The biggest advantage of flex volume is it’s simplicity. It is just an adapter around the internal api. We will keep it simple in the new design and try to follow the same adapter pattern.
+
+### Non Goals:
+* This proposal does not the address the poor installation model of existing model. It is covered in another PR #32550.
+
+## Terminology:
+**Plugin**: refers to flex volume plugin.
+
+**Plugin Driver API**: API calls defined by flex volume plugin.
+
+**Plugin Driver**: Out-of-tree driver/plugin which implements plugin driver api.
+
+**In-tree plugin**: Plugins which are included in Kubernetes project/source tree.
+
+## API:
+Existing flex volume plugin driver api is in alpha phase and are subject to change between Kubernetes versions. This section proposes stable api for plugin driver writers to work with.
+
+### Proposal:
+Proposes fewer APIs with stable request & response API objects.
+
+###### Pros:
+* Has standard request & response API objects.
+* Supports versioning.
+* Uses gRPC.
+* Fewer calls.
+* Simple and clean interface.
+
+###### Cons:
+* New API objects.
+
+#### APIs
+
+```protobuf
+
+// Capabilities of the driver
+message DriverCapabilities {
+	string supported_read_write_modes = 1;
+	enum AttachmentPolicy {
+	    NONE = 0;  // Driver does not support attach/detach.
+	    LOCAL = 1; // Driver supports local volume attachments from kubelet.
+	    REMOTE = 2; // Driver supports remote volume attachments to a node from controller-manager.
+	}
+	AttachmentPolicy supported_attachment_policy = 2;
+	bool supports_dynamic_provisioning = 3;
+	bool supports_custom_mount = 4;
+	bool supports_selinux = 5;
+	bool supports_ownership_management = 6;
+	bool supports_metrics = 7;
```

Comment:
Could these be consolidated into bit mask? I am concerned if new capabilities are added and an old drivers fail to handle them.